### ISSUE TITLE
Questionable package name validation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11800]
Migrated from [rt.perl.org#105922](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=105922) (status was 'resolved')

Searchable as RT105922$

**Comment by p5pRT** [comment]
### From @cpansprout

This is apparently a valid package name&#8203;:

$ perl \-le 'q"\_\#@&#8203;\*$\!@&#8203;\*^\("\->can\(can\); print \-ok'
\-ok

But this is not&#8203;:

$ perl \-le 'q"\#@&#8203;\*$\!@&#8203;\*^\(\_"\->can\(can\); print \-ok'
Can't call method "can" without a package or object reference at \-e line 1\.

That also means that main&#8203;::foo is not always equivalent to :&#8203;:foo&#8203;:

$ perl \-le 'main&#8203;::foo\->can\(can\); print \-ok'
\-ok
$ perl \-le '&#8203;::foo\->can\(can\); print \-ok'
Can't call method "can" without a package or object reference at \-e line 1\.

If we could eliminate that check\, it would allow for far more possibilities with UNIVERSAL methods\-\-a bit like autobox\.

I think this is a pity&#8203;:

$ \./perl \-Ilib \-le' print foo\->CORE&#8203;::uc'
FOO
$ \./perl \-Ilib \-le' print "3foo"\->CORE&#8203;::uc'
Can't call method "CORE&#8203;::uc" without a package or object reference at \-e line 1\.

Also\, the Unicode Bug affects class method calls\.  Fixing the above \(which would fix this too\) is easier than fixing just this&#8203;:

$ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; $p\->can\(can\)'

$ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; utf8&#8203;::downgrade $p; $p\->can\(can\)'
Can't call method "can" without a package or object reference at \-e line 1\.

It doesn’t even work when the package exists\.

And to drive the point home&#8203;:

$ perl \-e 'sub foo \{ warn $&#8203;::Moose&#8203;::VERSION \} use Moose'
Can't call method "can" without a package or object reference at /Library/Perl/5\.10\.1/darwin\-thread\-multi\-2level/Moose/Exporter\.pm line 348\.



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.15\.4&#8203;:

Configured by sprout at Wed Nov  2 09&#8203;:06&#8203;:14 PDT 2011\.

Summary of my perl5 \(revision 5 version 15 subversion 4\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: f3640611309ab8d6271598d071119f09fd9e8cf0
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.5\.0\, archname=darwin\-thread\-multi\-2level
&nbsp;    uname='darwin pint\.local 10\.5\.0 darwin kernel version 10\.5\.0&#8203;: fri nov 5 23&#8203;:20&#8203;:39 pdt 2010; root&#8203;:xnu\-1504\.9\.17~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Doptimize=\-g \-Dusedevel \-Duseithreads \-Dmad'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5664\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.15\.4&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.4/darwin\-thread\-multi\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.4
&nbsp;    /usr/local/lib/perl5/5\.15\.4/darwin\-thread\-multi\-2level
&nbsp;    /usr/local/lib/perl5/5\.15\.4
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.15\.4&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/sprout
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @ikegami

On Sun\, Dec 11\, 2011 at 4&#8203;:35 PM\, Father Chrysostomos \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> Also\, the Unicode Bug affects class method calls\.  Fixing the above \(which
> would fix this too\) is easier than fixing just this&#8203;:
>
> $ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; $p\->can\(can\)'
>
> $ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; utf8&#8203;::downgrade $p;
> $p\->can\(can\)'
> Can't call method "can" without a package or object reference at \-e line 1\.
>

And feature unicode\_strings doesn't help\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Hugmeir

On Tue\, Dec 13\, 2011 at 5&#8203;:48 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:

> On Sun\, Dec 11\, 2011 at 4&#8203;:35 PM\, Father Chrysostomos \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>> Also\, the Unicode Bug affects class method calls\.  Fixing the above
>> \(which would fix this too\) is easier than fixing just this&#8203;:
>>
>> $ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; $p\->can\(can\)'
>>
>> $ \./perl \-Ilib \-Mutf8 \-le '$p = "þackage"; utf8&#8203;::downgrade $p;
>> $p\->can\(can\)'
>> Can't call method "can" without a package or object reference at \-e line
>> 1\.
>>
>
> And feature unicode\_strings doesn't help\.
>
>
Whoops\, that was me\. Fixing this is one line in pp\_hot\.c \(patch attached\)\,
but that part of the code is just\.\. weird\.

A valid class name\, as far as the left side of \-> is concerned\, is either
something that is already in the stash cache hash\, and failing that\,
something that has a IO slot\, and failing that\, something that matches
/^\[\\p\{XIDS\}\_\]/\.
I think that explains why \_\#@&#8203;\*$\!@&#8203;\*^\( works\, but \#@&#8203;\*$\!@&#8203;\*^\(\_ doesn't\, and
similarly with :&#8203;:foo\. Though it sure seems\.\.\. quirky\.

The IO part is the real kicker though&#8203;: You can make \*anything\* a valid
class name by doing something like perl \-E 'open q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\, ">"\, undef;
q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\->can\(can\)'

**Comment by p5pRT** [comment]
### From @Hugmeir

<details><summary>0001-pp_hot.c-First-letter-of-latin-1-classnames-wasn-t.patch</summary>

```diff
From a9ad2aee8acc54e97fdd09e75ae9390b5b2807b2 Mon Sep 17 00:00:00 2001
From: Brian Fraser <fraserbn@gmail.com>
Date: Tue, 13 Dec 2011 19:32:48 -0300
Subject: [PATCH] pp_hot.c: First letter of latin-1 classnames wasn't being checked correctly.

Previously the first letter for latin-1 classnames was being mischecked, only
allowing ASCII, which caused an instance of the Unicode Bug for downgradable
classnames.
---
 pp_hot.c        |    2 +-
 t/uni/package.t |   19 ++++++++++++++++++-
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/pp_hot.c b/pp_hot.c
index 70eb5a1..a2d6f91 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -2973,7 +2973,7 @@ S_method_common(pTHX_ SV* meth, U32* hashp)
 	    if (!packname ||
 		((UTF8_IS_START(*packname) && DO_UTF8(sv))
 		    ? !isIDFIRST_utf8((U8*)packname)
-		    : !isIDFIRST(*packname)
+		    : !isIDFIRST_L1((U8)*packname)
 		))
 	    {
 		Perl_croak(aTHX_ "Can't call method \"%"SVf"\" %s",
diff --git a/t/uni/package.t b/t/uni/package.t
index 317ddd4..bb9092b 100644
--- a/t/uni/package.t
+++ b/t/uni/package.t
@@ -6,7 +6,7 @@ BEGIN {
     require './test.pl';
 }
 
-plan (tests => 16);
+plan (tests => 18);
 
 use utf8;
 use open qw( :utf8 :std );
@@ -92,3 +92,20 @@ ok 1, "sanity check. If we got this far, UTF-8 in package names is legal.";
     eval q[package ��� {];
     like $@, qr/\AMissing right curly /, "comp/package_block.t test";
 }
+
+# perl #105922
+
+{
+   my $latin_1 = "��ackage";
+   my $utf8    = "��ackage";
+   utf8::downgrade($latin_1);
+   utf8::upgrade($utf8);
+
+   local $@;
+   eval { $latin_1->can("yadda") };
+   ok(!$@, "latin1->meth works");
+
+   local $@;
+   eval { $utf8->can("yadda") };
+   ok(!$@, "utf8->meth works");
+}
-- 
1.5.6.3


```
</details>

**Comment by p5pRT** [comment]
### From tchrist@perl.com

> A valid class name\, as far as the left side of \-> is concerned\, is either
> something that is already in the stash cache hash\, and failing that\,
> something that has a IO slot\, and failing that\, something that matches
> /^\[\\p\{XIDS\}\_\]/\.

Oh darn it\.   I had convinced myself that Perl identifiers could 
also start with \\p\{PC\}\.  Or at least\, some of them can\, which 
I do know\.  It's just not clear which are which\.

And which ones can start with \\p\{Nd\} is also anything but clear\.

Not clear\, not clean\.

\-\-tom

**Comment by p5pRT** [comment]
### From @Hugmeir

On Tue\, Dec 13\, 2011 at 8&#8203;:12 PM\, Tom Christiansen \<tchrist@&#8203;perl\.com> wrote&#8203;:

> > A valid class name\, as far as the left side of \-> is concerned\, is either
> > something that is already in the stash cache hash\, and failing that\,
> > something that has a IO slot\, and failing that\, something that matches
> > /^\[\\p\{XIDS\}\_\]/\.
>
> Oh darn it\.   I had convinced myself that Perl identifiers could
> also start with \\p\{PC\}\.  Or at least\, some of them can\, which
> I do know\.  It's just not clear which are which\.
>
> And which ones can start with \\p\{Nd\} is also anything but clear\.
>
> Not clear\, not clean\.
>
>
That's rightish too\. But do note that I said class names on the left side
of \->\, not identifiers in general \( that can of worms is well known :\)\.
"1"\->can\(yadda\) won't work\, even though 1 is a valid identifier\, and so on\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Dec 13 14&#8203;:58&#8203;:38 2011\, Hugmeir wrote&#8203;:
> On Tue\, Dec 13\, 2011 at 5&#8203;:48 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:
> 
> > On Sun\, Dec 11\, 2011 at 4&#8203;:35 PM\, Father Chrysostomos \<
> > perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >
> >> Also\, the Unicode Bug affects class method calls\.  Fixing the above
> >> \(which would fix this too\) is easier than fixing just this&#8203;:
> >>
> >> $ \./perl \-Ilib \-Mutf8 \-le '$p = "�ackage"; $p\->can\(can\)'
> >>
> >> $ \./perl \-Ilib \-Mutf8 \-le '$p = "�ackage"; utf8&#8203;::downgrade $p;
> >> $p\->can\(can\)'
> >> Can't call method "can" without a package or object reference at \-e
line
> >> 1\.
> >>
> >
> > And feature unicode\_strings doesn't help\.
> >
> >
> Whoops\, that was me\. Fixing this is one line in pp\_hot\.c \(patch attached\)\,
> but that part of the code is just\.\. weird\.
> 
> A valid class name\, as far as the left side of \-> is concerned\, is either
> something that is already in the stash cache hash\, and failing that\,
> something that has a IO slot\, and failing that\, something that matches
> /^\[\\p\{XIDS\}\_\]/\.
> I think that explains why \_\#@&#8203;\*$\!@&#8203;\*^\( works\, but \#@&#8203;\*$\!@&#8203;\*^\(\_ doesn't\, and
> similarly with :&#8203;:foo\. Though it sure seems\.\.\. quirky\.
> 
> The IO part is the real kicker though&#8203;: You can make \*anything\* a valid
> class name by doing something like perl \-E 'open q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\, ">"\,
undef;
> q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\->can\(can\)'

That’s not a class method call\.  That’s an IO method\.

$ perl \-E 'open q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\, ">"\, undef; q\<\#@&#8203;\*$\!@&#8203;\*^\(\_>\->oeunt'
Can't locate object method "oeunt" via package "IO&#8203;::Handle" at \-e line 1\.

And 5\.9\.5 had a bug making constants turn into IOs&#8203;:

perl5\.9\.5 \-Mconstant=Just\,0 \-le 'sub
IO&#8203;::Handle&#8203;::a\_japh\{print\*\{$\_\[0\]\}=~/\(\[^&#8203;:\]\+\)$/\," another Perl hacker\,"\}
"Just"\->a\_japh'

:\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Dec 13 14&#8203;:58&#8203;:38 2011\, Hugmeir wrote&#8203;:
> On Tue\, Dec 13\, 2011 at 5&#8203;:48 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:
> 
> > On Sun\, Dec 11\, 2011 at 4&#8203;:35 PM\, Father Chrysostomos \<
> > perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >
> >> Also\, the Unicode Bug affects class method calls\.  Fixing the above
> >> \(which would fix this too\) is easier than fixing just this&#8203;:
> >>
> >> $ \./perl \-Ilib \-Mutf8 \-le '$p = "�ackage"; $p\->can\(can\)'
> >>
> >> $ \./perl \-Ilib \-Mutf8 \-le '$p = "�ackage"; utf8&#8203;::downgrade $p;
> >> $p\->can\(can\)'
> >> Can't call method "can" without a package or object reference at \-e
line
> >> 1\.
> >>
> >
> > And feature unicode\_strings doesn't help\.
> >
> >
> Whoops\, that was me\. Fixing this is one line in pp\_hot\.c \(patch attached\)\,
> but that part of the code is just\.\. weird\.

Thank you\.  Applied as d47f310d69\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @doy

So is the way to move forward here to just make any string valid on the
lhs of \->? I think this would make a lot more sense than the current
behavior\.

\-doy

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Jul 03 14&#8203;:04&#8203;:57 2012\, doy wrote&#8203;:
> So is the way to move forward here to just make any string valid on the
> lhs of \->? I think this would make a lot more sense than the current
> behavior\.

Yes\, I think so\.  It’s on my to\-do list\, but my to\-do list is \*long\*\. :\-\)

In other words\, I would be happy for someone to get to this before I do\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue Jul 03 14&#8203;:04&#8203;:57 2012\, doy wrote&#8203;:
> So is the way to move forward here to just make any string valid on the
> lhs of \->? I think this would make a lot more sense than the current
> behavior\.

Yes\, I think so\.  It’s on my to\-do list\, but my to\-do list is \*long\*\. :\-\)

In other words\, I would be happy for someone to get to this before I do\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

On Tue Jul 03 14&#8203;:04&#8203;:57 2012\, doy wrote&#8203;:
> So is the way to move forward here to just make any string valid on the
> lhs of \->? I think this would make a lot more sense than the current
> behavior\.

Yes\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 04 05&#8203;:46&#8203;:31 2012\, rjbs wrote&#8203;:
> On Tue Jul 03 14&#8203;:04&#8203;:57 2012\, doy wrote&#8203;:
> > So is the way to move forward here to just make any string valid on the
> > lhs of \->? I think this would make a lot more sense than the current
> > behavior\.
> 
> Yes\.

Test&#8203;::More fails its tests with the attachment\, because it’s doing this&#8203;:

&nbsp;            elsif\( $error =~ /Can't call method "isa" without a package/ \) \{

but "anything"\->isa\("Wibble"\) will simply return false now\, instead of
dying\.

That means Test&#8203;::More&#8203;::isa\_ok may output&#8203;:

\#   Failed test 'My Wibble isa Wibble'
\#   at t/fail\-more\.t line 248\.
\#     My Wibble isn't a 'Wibble' it's a ''

instead of what its tests are checking for&#8203;:

\#   Failed test 'My Wibble isa Wibble'
\#   at t/fail\-more\.t line 248\.
\#     My Wibble isn't a class or reference

\(That ‘it's a ''’ is another bug\, reported as \#78204\.\)

I think Test&#8203;::More’s tests will need a little bit of adjustment\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

From 8eff703c953b1daece21d3ef7bda95618fac1a0a Mon Sep 17 00&#8203;:00&#8203;:00 2001
From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;: Wed\, 4 Jul 2012 11&#8203;:25&#8203;:52 \-0700
Subject&#8203;: \[PATCH\] \[perl \#105922\] Allow any string before \->meth
MIME\-Version&#8203;: 1\.0
Content\-Type&#8203;: text/plain; charset=UTF\-8
Content\-Transfer\-Encoding&#8203;: 8bit

The rules for filtering out what do not look like package names are
not logical and disallow valid things like "&#8203;::main"\, while allowing
q"\_\#@&#8203;\*$\!@&#8203;\*^\("\.

This commit simply lets any non\-empty string be used as a package
name\.  If it is a typo\, you’ll get an error anyway\.  This allows
autobox\-style calls like "3foo"\->CORE&#8203;::uc\, or even "3foo"\->uc if you
set up @&#8203;ISA first\.

I made an exception for the empty string because it messes up caches
somehow and causes subsequent method calls all to be called on the
main package\.  I haven’t looked into that yet\.  I don’t know whether
it’s worth it\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/pp_hot.c b/pp_hot.c
index 77b707c..6366aee 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -2960,12 +2960,14 @@ S_method_common(pTHX_ SV* meth, U32* hashp)
     PERL_ARGS_ASSERT_METHOD_COMMON;
 
     if (!sv)
+       undefined:
 	Perl_croak(aTHX_ "Can't call method \"%"SVf"\" on an undefined value",
 		   SVfARG(meth));
 
     SvGETMAGIC(sv);
     if (SvROK(sv))
 	ob = MUTABLE_SV(SvRV(sv));
+    else if (!SvOK(sv)) goto undefined;
     else {
 	GV* iogv;
         STRLEN packlen;
@@ -2994,17 +2996,11 @@ S_method_common(pTHX_ SV* meth, U32* hashp)
 	    !(ob=MUTABLE_SV(GvIO(iogv))))
 	{
 	    /* this isn't the name of a filehandle either */
-	    if (!packname ||
-		((UTF8_IS_START(*packname) && DO_UTF8(sv))
-		    ? !isIDFIRST_utf8((U8*)packname)
-		    : !isIDFIRST_L1((U8)*packname)
-		))
+	    if (!packname || !packlen)
 	    {
-		/* diag_listed_as: Can't call method "%s" without a package or object reference */
-		Perl_croak(aTHX_ "Can't call method \"%"SVf"\" %s",
-			   SVfARG(meth),
-			   SvOK(sv) ? "without a package or object reference"
-				    : "on an undefined value");
+		Perl_croak(aTHX_ "Can't call method \"%"SVf"\" "
+				 "without a package or object reference",
+				  SVfARG(meth));
 	    }
 	    /* assume it's a package name */
 	    stash = gv_stashpvn(packname, packlen, packname_is_utf8 ? SVf_UTF8 : 0);
diff --git a/t/op/method.t b/t/op/method.t
index 09f6ee3..8f6bfb8 100644
--- a/t/op/method.t
+++ b/t/op/method.t
@@ -13,7 +13,7 @@ BEGIN {
 use strict;
 no warnings 'once';
 
-plan(tests => 98);
+plan(tests => 103);
 
 @A::ISA = 'B';
 @B::ISA = 'C';
@@ -417,3 +417,15 @@ eval { () = undef; new {} };
 like $@,
      qr/^Can't call method "new" without a package or object reference/,
     'Err msg from new{} when stack contains undef';
+
+sub flomp { "flimp" }
+sub main::::flomp { "flump" }
+is "::"->flomp, 'flump', 'method call on ::';
+is "::main"->flomp, 'flimp', 'method call on ::main';
+eval { ""->flomp };
+like $@,
+     qr/^Can't call method "flomp" without a package or object reference/,
+    'method call on empty string';
+is "3foo"->CORE::uc, '3FOO', '"3foo"->CORE::uc';
+{ no strict; @{"3foo::ISA"} = "CORE"; }
+is "3foo"->uc, '3FOO', '"3foo"->uc (autobox style!)';
diff --git a/t/run/fresh_perl.t b/t/run/fresh_perl.t
index cd5899a..376ceaf 100644
--- a/t/run/fresh_perl.t
+++ b/t/run/fresh_perl.t
@@ -81,7 +81,7 @@ $array[128]=1
 ########
 $x=0x0eabcd; print $x->ref;
 EXPECT
-Can't call method "ref" without a package or object reference at - line 1.
+Can't locate object method "ref" via package "961485" (perhaps you forgot to load "961485"?) at - line 1.
 ########
 chop ($str .= <DATA>);
 ########
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed Jul 04 05&#8203;:46&#8203;:31 2012\, rjbs wrote&#8203;:
> On Tue Jul 03 14&#8203;:04&#8203;:57 2012\, doy wrote&#8203;:
> > So is the way to move forward here to just make any string valid on the
> > lhs of \->? I think this would make a lot more sense than the current
> > behavior\.
> 
> Yes\.

Test&#8203;::More fails its tests with the attachment\, because it’s doing this&#8203;:

&nbsp;            elsif\( $error =~ /Can't call method "isa" without a package/ \) \{

but "anything"\->isa\("Wibble"\) will simply return false now\, instead of
dying\.

That means Test&#8203;::More&#8203;::isa\_ok may output&#8203;:

\#   Failed test 'My Wibble isa Wibble'
\#   at t/fail\-more\.t line 248\.
\#     My Wibble isn't a 'Wibble' it's a ''

instead of what its tests are checking for&#8203;:

\#   Failed test 'My Wibble isa Wibble'
\#   at t/fail\-more\.t line 248\.
\#     My Wibble isn't a class or reference

\(That ‘it's a ''’ is another bug\, reported as \#78204\.\)

I think Test&#8203;::More’s tests will need a little bit of adjustment\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jul 04 12&#8203;:36&#8203;:34 2012\, sprout wrote&#8203;:
> Test&#8203;::More fails its tests with the attachment\,

which is also on the sprout/package\-names\-113974 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed Jul 04 12&#8203;:36&#8203;:34 2012\, sprout wrote&#8203;:
> Test&#8203;::More fails its tests with the attachment\,

which is also on the sprout/package\-names\-113974 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-comment@&#8203;perl\.org> \[2012\-07\-04T15&#8203;:36&#8203;:35\]
> Test&#8203;::More fails its tests with the attachment\, because it’s doing this&#8203;:
> 
>             elsif\( $error =~ /Can't call method "isa" without a package/ \) \{
> 
> but "anything"\->isa\("Wibble"\) will simply return false now\, instead of
> dying\.
> \[…\]
> I think Test&#8203;::More’s tests will need a little bit of adjustment\.

I think so\, too\.

Schwern\, what say you?

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @schwern

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-
Hash&#8203;: SHA1

On 2012\.7\.5 7&#8203;:53 PM\, Ricardo Signes wrote&#8203;:
>> I think Test&#8203;::More’s tests will need a little bit of adjustment\.
> 
> I think so\, too\.
> 
> Schwern\, what say you?

I have no objection\.


\- \-\- 
24\. Must not tell any officer that I am smarter than they are\, especially
&nbsp;    if it's true\.
&nbsp;    \-\- The 213 Things Skippy Is No Longer Allowed To Do In The U\.S\. Army
&nbsp;           http&#8203;://skippyslist\.com/list/
\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: GnuPG v2\.0\.19 \(Darwin\)
Comment&#8203;: Using GnuPG with Mozilla \- http&#8203;://enigmail\.mozdev\.org/

iEYEARECAAYFAk/2WrgACgkQWMohlhD1QyfosgCfQRSIoGkdAK7JGRJre4eKalZY
7H4AoIgg0sMRxmkfj6XIaVw8WmIvZBGE
=9JuS
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in 7156e69ab\.
\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in 7156e69ab\.
\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
